pango/src/attribute.rs

```diff
@@ -200,4 +204,200 @@ impl Attribute {
             (*stash.0).end_index = index;
         }
     }
+
+    pub fn downcast<T: IsAttribute>(self) -> Result<T, Attribute> {
+        unsafe {
+            if T::ATTR_TYPES.contains(&self.get_attr_class().type_()) {
+                Ok(from_glib_full(self.to_glib_full()))
+            } else {
+                Err(self)
+            }
+        }
+    }
+
+    pub fn downcast_ref<T: IsAttribute>(&self) -> Option<&T> {
+        unsafe {
+            if T::ATTR_TYPES.contains(&self.get_attr_class().type_()) {
+                Some(&*(self as *const Attribute as *const T))
+            } else {
+                None
+            }
+        }
+    }
+}
+
+pub trait IsAttribute:
+    FromGlibPtrFull<*const ffi::PangoAttribute>
+    + FromGlibPtrFull<*mut ffi::PangoAttribute>
+    + std::convert::AsRef<crate::Attribute>
+    + 'static
+{
+    const ATTR_TYPES: &'static [AttrType];
+    fn upcast(self) -> Attribute;
+    fn upcast_ref(&self) -> &Attribute;
+}
+
+macro_rules! define_attribute_struct {
+    ($rust_type:ident, $ffi_type:path, $attr_types:expr) => {
+
+        #[cfg(any(feature = "v1_44", feature = "dox"))]
+        #[cfg_attr(feature = "dox", doc(cfg(feature = "v1_44")))]
+        glib::wrapper! {
+            #[derive(Debug, PartialOrd, Ord, Hash)]
+            pub struct $rust_type(Boxed<$ffi_type>);
+
+            match fn {
+                copy => |ptr| ffi::pango_attribute_copy(ptr as *const ffi::PangoAttribute) as *mut $ffi_type,
+                free => |ptr| ffi::pango_attribute_destroy(ptr as *mut ffi::PangoAttribute),
+                get_type => || ffi::pango_attribute_get_type(),
+            }
+        }
+
+        #[cfg(not(any(feature = "v1_44", feature = "dox")))]
+        glib::wrapper! {
+            #[derive(Debug, PartialOrd, Ord, Hash)]
```

Comment:
As clippy suggests, if you implement `PartialEq` / `Eq` manually, you must implement `Hash` manually too and it must be consistent with those manual `PartialEq` / `Eq` implementations (equal must mean same hash).

Reply:
I see the problem.  I basically copied this from `src/auto/attribute.rs`, so the same problem exists there already.

We have this nice equals function provided for us, but no hash function. It would be silly to try to match a Hash function to work on all the same values.

I think the only thing that makes sense is to not derive `Hash` or `PartialEq` / `Eq` and just provide the equals function as a method.

Reply:
Maybe don't derive `Hash` then, and if someone needs it they can figure out a way around it.

Reply:
Why did you remove the `PartialEq` / `Eq` implementation around `pango_attribute_equal()` together with the others? IMHO that one should stay, `Hash` / `PartialOrd` / `Ord` should go away.